packages/teams-js/src/private/externalAppAuthentication.ts

```diff
@@ -0,0 +1,329 @@
+    requestType: OriginalRequestType.QueryMessageExtensionRequest;
+    commandId: string;
+    parameters?: {
+      name?: string;
+      value?: string;
+    }[];
+    queryOptions?: {
+      count: number;
+      skip: number;
+    };
+  }
+
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  interface IActionExecuteInvokeRequest {
+    requestType: OriginalRequestType.ActionExecuteInvokeRequest;
+    type: string; // "invoke"
+    id: string; // "action id"
+    verb: string; // "action"
+    data: Record<string, unknown>; //object; // {}
+  }
+
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  export enum OriginalRequestType {
+    ActionExecuteInvokeRequest = 'ActionExecuteInvokeRequest',
+    QueryMessageExtensionRequest = 'QueryMessageExtensionRequest',
+  }
+  /*********** END REQUEST TYPE ************/
+
+  /*********** BEGIN RESPONSE TYPE ************/
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  export type IInvokeResponse = IQueryMessageExtensionResponse | IActionExecuteResponse;
+
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  export enum InvokeResponseType {
+    ActionExecuteInvokeResponse = 'ActionExecuteInvokeResponse',
+    QueryMessageExtensionResponse = 'QueryMessageExtensionResponse',
+  }
+
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  export interface IQueryMessageExtensionResponse {
+    responseType: InvokeResponseType.QueryMessageExtensionResponse;
+    composeExtension?: ComposeExtensionResponse;
+  }
+
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  export interface IActionExecuteResponse {
+    responseType: InvokeResponseType.ActionExecuteInvokeResponse;
+    value: Record<string, unknown>;
+    signature?: string;
+    statusCode: number;
+    type: string;
+  }
+
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  export type ComposeExtensionResponse = {
+    attachmentLayout: AttachmentLayout;
+    type: ComposeResultTypes;
+    attachments: QueryMessageExtensionAttachment[];
+    suggestedActions?: QueryMessageExtensionSuggestedActions;
+    text?: string;
+  };
+
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  export type QueryMessageExtensionSuggestedActions = {
+    actions?: Actions[];
+  };
+
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  export type Actions = {
+    type: string;
+    title: string;
+    value: string;
+  };
+
+  /**
+   * @hidden
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  export type QueryMessageExtensionCard = {
+    contentType: string;
+    content: Record<string, unknown>; //object;
+    name?: string;
+    thumbnailUrl?: string;
```

Comment:
>string [](http://example.com/codeflow?start=19&length=6)

Should this be a `URL`? Same question for `contentUrl`.

Note that if you do switch it, it would likely need an intentional function to serialize the value if it gets sent over to the host SDK.

Reply:
No more URLs!